sklearn/_fast_gradient_boosting/gradient_boosting.py

```diff
@@ -0,0 +1,757 @@
+from .types import Y_DTYPE, X_DTYPE, X_BINNED_DTYPE
+
+from .binning import BinMapper
+from .grower import TreeGrower
+from .loss import _LOSSES
+
+
+class BaseFastGradientBoosting(BaseEstimator, ABC):
+    """Base class for fast gradient boosting estimators."""
+
+    @abstractmethod
+    def __init__(self, loss, learning_rate, n_estimators, max_leaf_nodes,
+                 max_depth, min_samples_leaf, l2_regularization, max_bins,
+                 scoring, validation_fraction, n_iter_no_change, tol, verbose,
+                 random_state):
+        self.loss = loss
+        self.learning_rate = learning_rate
+        self.n_estimators = n_estimators
+        self.max_leaf_nodes = max_leaf_nodes
+        self.max_depth = max_depth
+        self.min_samples_leaf = min_samples_leaf
+        self.l2_regularization = l2_regularization
+        self.max_bins = max_bins
+        self.n_iter_no_change = n_iter_no_change
+        self.validation_fraction = validation_fraction
+        self.scoring = scoring
+        self.tol = tol
+        self.verbose = verbose
+        self.random_state = random_state
+
+    def _validate_parameters(self):
+        """Validate parameters passed to __init__.
+
+        The parameters that are directly passed to the grower are checked in
+        TreeGrower."""
+
+        if self.loss not in self._VALID_LOSSES:
+            raise ValueError(
+                "Loss {} is not supported for {}. Accepted losses: "
+                "{}.".format(self.loss, self.__class__.__name__,
+                             ', '.join(self._VALID_LOSSES)))
+
+        if self.learning_rate <= 0:
+            raise ValueError('learning_rate={} must '
+                             'be strictly positive'.format(self.learning_rate))
+        if self.n_estimators < 1:
+            raise ValueError('n_estimators={} must not be smaller '
+                             'than 1.'.format(self.n_estimators))
+        if self.n_iter_no_change is not None and self.n_iter_no_change < 0:
+            raise ValueError('n_iter_no_change={} must be '
+                             'positive.'.format(self.n_iter_no_change))
+        if (self.validation_fraction is not None and
+                self.validation_fraction <= 0):
+            raise ValueError(
+                'validation_fraction={} must be strictly '
+                'positive, or None.'.format(self.validation_fraction))
+        if self.tol is not None and self.tol < 0:
+            raise ValueError('tol={} '
+                             'must not be smaller than 0.'.format(self.tol))
+
+    def fit(self, X, y):
+        """Fit the gradient boosting model.
+
+        Parameters
+        ----------
+        X : array-like, shape=(n_samples, n_features)
+            The input samples.
+
+        y : array-like, shape=(n_samples,)
+            Target values.
+
+        Returns
+        -------
+        self : object
+        """
+
+        fit_start_time = time()
```

Comment:
https://github.com/scikit-learn/scikit-learn/pull/11364#discussion_r227660418 states:

Use,

```python
start = timeit.default_timer()
```

for better precision on Windows.